feat: migrate from rebase to refresh terminology

HonkingGoose opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->

## Changes:

<!-- Describe what this pull request changes. -->

This was a relatively straightforward docs change from `rebase` to `refresh`.

But it's now a feature to replace `rebase` with `refresh` in code and docs.

## Context:

<!-- Describe why you're making these changes if it's not already explained in a corresponding issue. -->
<!-- If you're closing an existing issue with this pull request, use the keyword Closes #issue_number -->

I find the `rebase` terminology confusing, and want to replace it with something else.

Merging this PR helps with issue #7626.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added unit tests, or
- [ ] Unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/master/.github/pull_request_template.md -->

<!-- Please do not force push to this PR's branch after you have created this PR, as doing so makes it harder for us to review your work. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

HonkingGoose commented:
I'm not sure that the Renovate "rebase" is functionally different from a Git rebase.

## Starting point

Say a user has this Git history:
```git
$ git log
1111111 create renovate.json
0000000 initial commit with package.json file
```

## Renovate creates a branch to update a dependency:

```git
$ git log --graph --all --decorate

* 2222222 (renovate-update) fix(deps): update express to 4.17.1
* 1111111 (main) create renovate.json
* 0000000 initial commit with package.json file
```

## User commits to `main` instead of merging the update first:

```git
$ git commit -m "fix bug"

$ git log --graph --all --decorate

* 3333333 (main) fix bug
| * 2222222 (renovate-update) fix(deps): update express to 4.17.1
|/
* 1111111 create renovate.json
* 0000000 initial commit with package.json file
```

## User does Git rebase:

A `git rebase` basically throws away commit 2222222 and creates a new commit (4444444) that points to 3333333:

```git
$ git branch

On branch renovate-update

$ git rebase main
Throw away commit 2222222, and replace it with 4444444 which points to 3333333 on main.

$ git log --graph --all --decorate

* 4444444 (renovate-update) fix(deps): update express to 4.17.1
* 3333333 (main) fix bug
* 1111111 create renovate.json
* 0000000 initial commit with package.json file

$ git switch main
$ git merge renovate-update

Fast forward to 4444444.
```

## Renovate bot does "rebase"

Renovate also "throws away" commit 2222222, and replaces it with a new commit 4444444 that points to commit 3333333 on main. I don't think it matters that the Renovate back-end is not literally doing a `git rebase` but instead just does a completely fresh rerun which clobbers over the branch.

## Conclusion:

I think we can scrap the message that explains that Renovate does its own version of Git rebase.
Is this really helpful for readers? The end result is the same as far as the Git history is concerned:
- `git rebase` throws away original commit in favor of a new one pointing to current `main`
- Renovate "rebase": clobbers over original commit with a new run, which creates a new commit that points to current `main`

HonkingGoose commented on docs/usage/updating-rebasing.md:
```diff
@@ -1,53 +1,62 @@
+- When a newer version of the dependency is released
+- When you request a manual rebase from the Renovate bot
+
+Renovate uses its own version of "rebasing", which is _not the same_ as doing a `git rebase` with Git.
+Instead of doing a `git rebase`, Renovate manually recreates the PR commit based off of the latest commit in the base branch.
```
```suggestion
```

Scrap this section see https://github.com/renovatebot/renovate/pull/8029#issuecomment-745314007 for my reasoning.

rarkins commented:
The rebase is definitely not a real rebase, and it's important that we are technically accurate. The use case you describer is correct, but that's the simple case where nothing changed and nothing is conflicted. There are other use cases:

1. Current version is `1.0.0`, the PR has `1.0.1`, and `1.0.2` is released. In this case we "rebase" to update to `1.0.2`
2. A Renovate PR is merged, causing a conflict with another PR. In this case the branch is completely re-calculated when it is "rebased". There may be multiple different package resolutions as a result of the initial PR being merged.
3. The PR had an artifact when created, so the user retries/rebases and then this time the artifact generation works and additional files are added to the PR

In all the above, the result is not the same as a git rebase, but we use the term "rebase".

HonkingGoose commented:
> The rebase is definitely not a real rebase, and it's important that we are technically accurate.

I agree, we should be technically accurate, which is the best kind of accurate. 🙃 

> In all the above, the result is not the same as a git rebase, but we use the term "rebase".

I think my main problem with the use of the word "rebase" is that it's already a locked-in term, if you search for "rebase" you'll only find Git rebase.
I think it's a bit confusing to use the word "rebase" when we're not really talking about the Git way of rebasing.

How about we use a different word instead of "rebase"? That way we don't have to explain that it's different to a Git rebase, because we don't even use the word rebase. 😄 I also think this will make the text a bit easier to read, as you can't get confused with the way Git rebases.

Maybe we can use a word like:
- recreate
- refresh
- regenerate
- rejuvenate
- remake
- do-over

I think those words are closer to the mark of what Renovate is actually doing?
I don't know if this is a good or bad idea, so I'm open for your feedback on this. 👍

JamieMagee commented:
I agree that using the word "rebase" is confusing, and I have had to explain this in the past to people I've been helping to setup renovate.

I am a fan of "recreate" as a replacement for "rebase" 👍

HonkingGoose commented:
I'll go make a commit then, which changes all instances of "rebase" with "recreate", so that we can see how that looks.
I'll keep it to a single commit, so we can easily revert back out with `git revert` if we don't like my change. 😄

HonkingGoose commented:
@JamieMagee I've made a commit which changes all instances of "rebase" to recreate.

If we really want to migrate from the word "rebase" to "recreate" we need to:
- [ ] Rename the file I'm targeting with this PR.
- [ ] Check for broken links due to the renaming of the file.
- [ ] Add a migration to migrate users to `recreateWhen` instead of `rebaseWhen`.
- [ ] Adjust the bot template PR, so that it uses the proper term for it's "rebase/retry" checkbox.

HonkingGoose commented:
@rarkins what do you think about my idea to replace "rebase" with "recreate" in the docs and in the Renovate codebase? See https://github.com/renovatebot/renovate/pull/8029#issuecomment-757130021 for the full comment.

rarkins commented:
"recreate" doesn't quite sit right with me (yet). We could also consider "re-commit" (not a real term..) or "reapply".

Ideally I want to avoid people thinking they'll get a *new PR* so anything that can subtly imply it's mostly a change to the branch/commit would be best.

One downside of all terms apart from the `rebase` one is that it's not intuitive that the branch will also be brought up-to-date with the base branch.

The full term would probably be "rebase and reapply". i.e. rebase the branch and reapply the updates.

HonkingGoose commented:
> "recreate" doesn't quite sit right with me (yet). We could also consider "re-commit" (not a real term..) or "reapply".

Have you seen my list of alternative terms? Did you not like any of those? (I'm not clear if you've seen that list already...)

* refresh
* regenerate
* rejuvenate
* remake
* do-over

> Ideally I want to avoid people thinking they'll get a _new PR_ so anything that can subtly imply it's mostly a change to the branch/commit would be best.

Agreed.

> One downside of all terms apart from the `rebase` one is that it's not intuitive that the branch will also be brought up-to-date with the base branch.

Hmm, the term `rebase` does indeed imply getting up-to-date with the base branch.
So that's one point in favor of the `rebase` term 😄 

> The full term would probably be "rebase and reapply". i.e. rebase the branch and reapply the updates.

I like that this more closely matches the real backend behavior, but I dislike that we're then still using the "rebase" term in a way that's not really the Git rebase.

rarkins commented:
I'm leaning towards `refresh`. It's short, and it's not "misleading".

@JamieMagee @viceice

HonkingGoose commented:
I have reverted the first example where I changed `rebase` into `recreate`.

I have pushed a commit which changes `rebase` into `refresh` instead, as **rarkins** proposed.
This way we can get a feel for how we like `refresh` as a term.

rarkins commented on docs/usage/updating-rebasing.md:
```diff
@@ -1,53 +1,59 @@
+# Refreshing branches
 
-There are many cases where Renovate will need to update a branch/PR after its initial creation, and this document will attempt to describe them.
+There are many situations in which Renovate must refresh a branch.
```
```suggestion
In Renovate terminology, _refreshing_ a branch means Renovate will re-apply all dependency updates in the branch on top of the latest commit from the base branch (e.g. `main`/`master`).
There are many situations in which Renovate might refresh a branch.
```

rarkins commented on docs/usage/updating-rebasing.md:
```diff
@@ -1,53 +1,59 @@
+## Newer dependency versions
 
-If an existing PR is open to upgrade dependency "foo" to v1.1.0 and then v1.1.1 is released, then Renovate will regenerate the branch again.
+If you have a existing PR to upgrade dependency "foo" to 1.1.0, and later 1.1.1 is released, then Renovate will regenerate the branch.
```
```suggestion
If you have an existing PR but even newer versions exist of a dependency in the PR then Renovate will refresh the branch to apply the newer version.
```

rarkins commented on docs/usage/updating-rebasing.md:
```diff
@@ -1,53 +1,59 @@
-It is possible to manually request that Renovate rebase a PR by ticking a rebase/retry checkbox on GitHub or GitLab, or by adding the label "rebase" to a Renovate PR.
-This label name is also configurable via the `rebaseLabel` config option too.
+You can request that Renovate refresh a PR by ticking the refresh checkbox on GitHub or Gitlab.
+Or you can add a "refresh" label to the PR.
```
This is a problem because people may be used to using the `rebase` label. Maybe we need to support both?

HonkingGoose replied:
I'd like to just migrate everybody to the new label and make a clean break with the old term.
But this is more work to program, more error prone, and causes noise for existing users.

So, I think we have no other option but to support both labels.
We can however only mention the newer labels, so that new users are "nudged" towards using the newer labels.

rarkins commented on docs/usage/updating-rebasing.md:
```diff
@@ -1,53 +1,59 @@
 
-It is possible to manually request that Renovate rebase a PR by ticking a rebase/retry checkbox on GitHub or GitLab, or by adding the label "rebase" to a Renovate PR.
-This label name is also configurable via the `rebaseLabel` config option too.
+You can request that Renovate refresh a PR by ticking the refresh checkbox on GitHub or Gitlab.
```
```suggestion
You can request that Renovate refresh a PR by ticking the refresh checkbox on GitHub or Gitlab.

You can instead request a PR refresh by renaming the PR title to begin with `refresh!`.
```

We need to think whether to support `rebase!` too.

HonkingGoose replied:
I'd like to just migrate everybody to the new term and make a clean break with the old term.
But this is more work to program, more error prone, and causes noise for existing users.

So, I think we have no other option but to support both terms.
We can however only mention the newer term in the docs, so that new users are "nudged" towards using the newer `refresh!` term in their PR titles.

rarkins reviewed (changes requested):
This needs to be accompanied with code migrations as well as probably other docs updates.

rarkins commented:
As mentioned in my last review, this needs to be more than a `docs:` PR - needs code changes and migration in addition to other docs updated too.

HonkingGoose commented:
> This needs to be accompanied with code migrations as well as probably other docs updates.

I think this is what we need to do to get this landed?
Feel free to add any missing items to my checklist by editing it.

## Checklist for migration towards `refresh`

- [ ] Rename the file I'm targeting with this PR to something appropriate.
- [ ] Check for broken links due to the renaming of the file.
- [ ] Add a migration to migrate users to `refreshWhen` instead of `rebaseWhen`.
- [ ] Adjust the bot template PR, so that it uses the proper term for it's "rebase/retry" checkbox, should be `refresh`.
- [ ] Fix all other documentation that still mentions "rebase" when it's actually "recreate".
- [ ] Change PR title from `docs: improve updating and rebasing branches documentation` to `feat: use term recreate instead of rebase`
- [ ] Adjust function names that use `rebase` to use `refresh` instead + fix unit-test(s) that rely on old function name.

HonkingGoose commented:
I'm `status:blocked` on this, because I don't know what changes I should make to the code or tests to get this feature landed.

I've searched trough the code for hits for `rebase` and found some places where I think the code should probably be changed.
But I don't understand how all the loose parts fit together, so I can't really make changes myself without breaking everything...

So if we want to land this feature anytime soon, I'll need help from somebody who does understand all this stuff. 😄

Chanyankit commented:
We have a phone connect  at-
February 1st, 3:00pm!
And my name is Chan.
I have to say.. My English is not really good, and I can't make sure do we
make each other clear..
And I afraid my phone have been hack by other who I don't know and without
my permission....because of that when you are talking to me by phone call
is so
Strange in my side...

And the Investment that should be $300 is truly happened.. And I still want
to Cooperation with you....but I need more details by mail...

Chan


在 2020年12月15日 週二 21:50，HonkingGoose <notifications@github.com> 寫道：

> Changes:
>
> Another big rewrite.
>
>    - Reduce capitalization in page title and headings
>    - Fix whitespace in ordered list
>    - Let Markdown deal with numbering the ordered list
>    - Major rewrite
>
> Context:
>
> Helps with issue #7626
> <https://github.com/renovatebot/renovate/issues/7626>.
> Documentation (please check one with an [x])
>
>    - I have updated the documentation, or
>    - No documentation update is required
>
> How I've tested my work (please tick one)
>
> I have verified these changes via:
>
>    - Code inspection only, or
>    - Newly added unit tests, or
>    - Unit tests + ran on a real repository
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/renovatebot/renovate/pull/8029
> Commit Summary
>
>    - reduce capitalization
>    - fix whitespace ordered list
>    - let Markdown renderer deal with numbering
>    - rewrite
>
> File Changes
>
>    - *M* docs/usage/updating-rebasing.md
>    <https://github.com/renovatebot/renovate/pull/8029/files#diff-67cb67c449c348a46b6c233e701df64846cc8ba12a3b0dc0b026876e872694da>
>    (51)
>
> Patch Links:
>
>    - https://github.com/renovatebot/renovate/pull/8029.patch
>    - https://github.com/renovatebot/renovate/pull/8029.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/renovatebot/renovate/pull/8029>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQQ2EXJXSHDVNNVMSOMXORTSU5SS7ANCNFSM4U4LYXUA>
> .
>

HonkingGoose commented:
@rarkins

I suspect this PR will go nowhere without outside help. I think it's not nice of me to rely on your team to fix this for me, so I'm thinking of closing this PR and starting from scratch with just the basic docs changes (not changing the terminology of rebase).

I can create a new issue to discuss the `feat:` changes we would need to rename the `rebaseWhen` to `refreshWhen` and other similar changes, so that a competent programmer can pick it up.

Can you let me know what you want me to do here?

rarkins commented:
@HonkingGoose that sounds good. It would be a pretty deep change

HonkingGoose commented:
Closing this PR to try again as a simpler docs only improvement which keeps the current `rebase` terms (for now).
Thank you everybody for your valuable input!

I have opened issue #9020 to discuss the full scope and changes needed to do a proper migration from `rebase` to `refresh` in code and docs.
